### PR TITLE
[Feature] FilterCommand

### PR DESCRIPTION
**Pull request by sp4ce-cowboy**
### Major Changes
- FilterCommand added
- Local UserGuide functionality (and the pdf) removed (and swapped in place with the `summary` handler)

### Minor Changes
- logic.commands
    - FilterCommand added
    - FindCommand: Fix MESSAGE_USAGE 
    - Clear, Exit, Reset, Help: Added equals method and/or toString methods
- logic.parser
    - FilterCommandParser added
    - Clear and Reset Parsers added (Exit parser removed)
    - Get, Help, List: Added equals method and/or toString methods
- model.transaction.predicates
    - Added TransactionContainsAllKeywordsPredicate
    - Renamed the previous combo predicate to TransactionContainsAnyKeywordsPredicate
    - DateTimePredicate: Add explantory comment
- ui
    - Remove userguide handler in MainWindow (and pdf file)
    - In-place substitute with Summary
- Tests
    - Tests for every single one of the above mentioned changes 

Screenshot 2023-10-30 at 8 25 09 PM

### CODE REVIEW

File: src/main/java/unicash/logic/parser/ClearTransactionsCommandParser.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof ClearTransactionsCommandParser)) {
+            return false;
+        }
+
+        return this instanceof ClearTransactionsCommandParser;
```
Comment (woojiahao) [inline src/main/java/unicash/logic/parser/ClearTransactionsCommandParser.java:39]: See `ClearTransactionsCommand`

File: src/main/java/unicash/logic/parser/ExitCommandParser.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+public class ExitCommandParser implements Parser<ExitCommand> {
+
+    /**
+     * Parses {@code userInput} into an {@code ExitCommand}and returns it.
+     *
+     * @throws ParseException if {@code userInput} does not conform the expected format
+     */
+    @Override
+    public ExitCommand parse(String userInput) throws ParseException {
+        assert userInput != null : "userInput cannot be null";
```
Comment (woojiahao) [inline src/main/java/unicash/logic/parser/ExitCommandParser.java:19]: Same question for using `asser` over `requireNonNull`

File: src/main/java/unicash/logic/parser/ExitCommandParser.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof ExitCommandParser)) {
+            return false;
+        }
+
+        return this instanceof ExitCommandParser;
```
Comment (woojiahao) [inline src/main/java/unicash/logic/parser/ExitCommandParser.java:39]: See `ClearTransactionCommand`

File: src/main/java/unicash/logic/parser/FilterCommandParser.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+public class FilterCommandParser implements Parser<FilterCommand> {
+
+    private TransactionContainsAllKeywordsPredicate filterPredicate;
+
+    /**
+     * Creates a {@code FilterCommandParser} object with a default
+     * {@code TransactionContainsAllKeywordsPredicate} predicate.
+     */
+    FilterCommandParser() {
+        this.filterPredicate = new TransactionContainsAllKeywordsPredicate();
```
Comment (woojiahao) [inline src/main/java/unicash/logic/parser/FilterCommandParser.java:38]: Can omit use of `this`

Comment (woojiahao) [reply src/main/java/unicash/logic/parser/FilterCommandParser.java:38]: Also I realised that you could also just initialise the `filterPredicate` where it is declared and skip the declaration of the constructor entirely

File: src/main/java/unicash/logic/parser/FilterCommandParser.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                ArgumentTokenizer.tokenize(args, PREFIX_NAME, PREFIX_TYPE, PREFIX_AMOUNT, PREFIX_DATETIME,
+                        PREFIX_CATEGORY, PREFIX_LOCATION);
+
+        String trimmedArgs = args.trim();
+        if (trimmedArgs.isEmpty() || !argMultimap.getPreamble().isEmpty()) {
+            throw new ParseException(
+                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, FilterCommand.MESSAGE_USAGE));
+        }
+
+        // TODO: Propose an future implementation with more than 2 types of transactions in DG
```
Comment (woojiahao) [inline src/main/java/unicash/logic/parser/FilterCommandParser.java:59]: Hmm, not sure if this is needed since tP is coming to a close soon right? Maybe good to include in the docs for @lipwei1808 who made the `AddTransactionCommand` to add your rationale for not hardcoding the type of transaction

File: src/main/java/unicash/logic/parser/FilterCommandParser.java
Comment (woojiahao) [inline src/main/java/unicash/logic/parser/FilterCommandParser.java:1]: Similar comment to the `FilterCommand` which is the reason for allowing variadic arguments for all fields, I would imagine that filtering would usually be done for a single filter per attribute

File: src/main/java/unicash/logic/parser/GetCommandParser.java [hunk 6]
```diff
@@ -30,5 +31,26 @@ public GetCommand parse(String args) throws ParseException {
...
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof GetCommandParser)) {
+            return false;
+        }
+
+        return this instanceof GetCommandParser;
```
Comment (woojiahao) [inline src/main/java/unicash/logic/parser/GetCommandParser.java:45]: See `ClearTransactionsCommand`

File: src/main/java/unicash/logic/parser/ListCommandParser.java [hunk 7]
```diff
@@ -22,4 +24,27 @@ public ListCommand parse(String userInput) throws ParseException {
...
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof ListCommandParser)) {
+            return false;
+        }
+
+        return this instanceof ListCommandParser;
```
Comment (woojiahao) [inline src/main/java/unicash/logic/parser/ListCommandParser.java:39]: See `ClearTransactionsCommand`

File: src/main/java/unicash/logic/parser/ResetCommandParser.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof ResetCommandParser)) {
+            return false;
+        }
+
+        return this instanceof ResetCommandParser;
```
Comment (woojiahao) [inline src/main/java/unicash/logic/parser/ResetCommandParser.java:39]: See `ClearTransactionsCommand`

File: src/main/java/unicash/logic/parser/UniCashParser.java [hunk 9]
```diff
@@ -77,13 +78,13 @@ public Command parseCommand(String userInput) throws ParseException {
             return new GetTotalExpenditureCommandParser().parse(arguments);
 
         case ClearTransactionsCommand.COMMAND_WORD:
-            return new ClearTransactionsCommand();
+            return new ClearTransactionsCommandParser().parse(arguments);
```
Comment (woojiahao) [inline src/main/java/unicash/logic/parser/UniCashParser.java:81]: Was wondering why is there new parsers for these commands and a new restriction to have no user input at all? I think `Parsers` are only apt when there's arguments to process, if there aren't any, then we don't need to impose restrictions or create new `Parsers` that just effectively create a new `Command` that we had already done

File: src/main/java/unicash/model/transaction/predicates/TransactionContainsAllKeywordsPredicate.java [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+ * </p> Encapsulated within is a list of transaction predicates, and this list can be accessed
+ * publicly, and modified. The test method returns true only if the input Transaction matches
+ * all predicates in this list.
+ */
+public class TransactionContainsAllKeywordsPredicate implements Predicate<Transaction> {
+
+    private List<Predicate<Transaction>> predicateList;
+
+    public TransactionContainsAllKeywordsPredicate() {
+        this.predicateList = new ArrayList<>();
```
Comment (woojiahao) [inline src/main/java/unicash/model/transaction/predicates/TransactionContainsAllKeywordsPredicate.java:31]: Can move initialisation to where the variable is declared

File: src/main/java/unicash/model/transaction/predicates/TransactionContainsAllKeywordsPredicate.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    public TransactionContainsAllKeywordsPredicate(List<Predicate<Transaction>> predicateList) {
+        this.predicateList = predicateList;
+    }
```
Comment (woojiahao) [inline src/main/java/unicash/model/transaction/predicates/TransactionContainsAllKeywordsPredicate.java:36]: Alternatively can consider using the `this()` call to initialize to another constructor

**Comment by sp4ce-cowboy** [comment]
> Was wondering why is there new parsers for these commands and a new restriction to have no user input at all? I think `Parsers` are only apt when there's arguments to process, if there aren't any, then we don't need to impose restrictions or create new `Parsers` that just effectively create a new `Command` that we had already done

@woojiahao Thanks for taking the time to write the detailed comments, but before all the other relatively minor issues I would like to address this first as I feel that it is quite important. Some time earlier I had added this (image at the bottom) to the end of our docs, and so this is something that I have been considering implementing since [this](https://github.com/AY2324S1-CS2103-T16-3/tp/pull/90) PR. 

While I agree with you that Parsers are only "apt" when there's arguments to process, and while it may make logical sense, the idea here is to not depend on what a potential user (read: tester) might consider logical and fashion the input restriction in such a way that not even something as minor as `help  1` could be flagged as a `bug.severityLow` or a `featureFlaw` - where a proposed feature is not "as perfectly implemented" as it can be, even though I agree that it doesn't really make sense to overengineer something like this in the first place.

The parsers here are not intended to be parsers that supplement their associated `Command` like the others, rather they are a defensive practice intended to prevent the user (again, read: tester) from entering any such commands, even if it may not seemingly affect the normal course of operation. Having parsers for every single command is not just a way to strictly control user input (whether that input is allowed or not is irrelevant), but in doing so, preventing any possible "unintended" uses of these commands, if so in any clever way that it might seem like a genuine flaw. 

Something like `help 1` could be a natural user input when they see the ordered list of available commands on the Results Display, and something like `clear_Transactions today` and `list 18-08-2023` are all reasonable cases for misunderstandings, and thus viable to be flagged as a bug. And those are just the things I can think of at the moment. 

Thus, just to be safe, I propose that we do create parser classes for all (not just the ones mentioned below) and explain in our UG/DG that our commands are intended to be used in these specific ways so that the (new) user understands perfectly what each command is supposed to be. In the below screenshot, I initially was concerned about overzealous input validation, but that actually applies more to the arguments and their formatting themselves rather than the command words, for which strict validation is actually more appropriate given that as opposed to arguments, command words precede some form of actual execution and loose checks on these could constitute a flaw, a bug, or just poor engineering in general, at least in the context of the module.

So tl;dr to prevent against _any_ kind of possible "bugs" even if those are only considered bugs from the POV of a few. 

image


Technically speaking though, while this "parsing" could be done within the Command itself, it seemed like a very ad-hoc and inelegant design so I just wrote classes for them, like all the other Commands.

File: src/main/java/unicash/logic/commands/FilterCommand.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+                            + "\n\nMultiple keywords for the same property is allowed, except for Transaction "
+                            + "Type for which there can only be one keyword. A minimum of one keyword "
+                            + "must be provided."
+            )
+            .addParameter(PREFIX_NAME, "Name", true, true)
+            .addParameter(PREFIX_TYPE, "Type", true, false)
+            .addParameter(PREFIX_AMOUNT, "Amount", true, true)
+            .addParameter(PREFIX_DATETIME, "DateTime", true, true)
+            .addParameter(PREFIX_LOCATION, "Location", true, true)
+            .addParameter(PREFIX_CATEGORY, "Category", true, true)
```
[the remark this answers is not in the document]
Comment (sp4ce-cowboy) [reply src/main/java/unicash/logic/commands/FilterCommand.java:40]: Actually this is the exact inverse of Find, and is intended to be so. If you look at the code, Find anyMatches everything and Filter only matches if everything is matched.

`find friends mcdonalds` would give me every transaction with friends as the name, category, location (e.g. _The Wine & Gourmet Friends_) and every transaction with mcdonalds as the name ... and so on.

`filter n/friends n/mcdonalds` allows the user to exactly narrow down to transactions that must contain **both** name (e.g. _Mcdonalds with friends)_. or `filter n/friends n/mcdonalds l/COM1` that must contain all parameters in order for it to match, so the user can find exactly what they are looking for, regardless of how precise. 

While adding more parameters to `find` increases its matches expansively, adding more parameters to `filter` decreases its matches conversely. I had actually originally implemented `filter` with each parameter non-variadic, but doing it this way just seems to make _sense_ somehow. It's like, if `find` can so broad then `filter` _should_ be as narrowable as possible, else it seems to be a bit asymmetric, and I think this find/filter synergy is a solid representation of the product design criteria (_...how well your features fit together to form a cohesive product_).

Either way though, the change is a simple one line expansion of the `verifyNoDuplicatePrefixesFor` method inside FilterCommandParser. But I really think that the find/filter contrast is like the _ideal_ definition of feature-cohesion, and all we would have to do is explain this in the UG/DG.

File: src/main/java/unicash/logic/parser/FilterCommandParser.java (hunk 4 above)
Comment (sp4ce-cowboy) [reply src/main/java/unicash/logic/parser/FilterCommandParser.java:38]: Correct me if I'm wrong, but I believe explicit constructor declaration and readability is more in alignment with defensive programming principles. I'm not sure where I read that in last week's content, but I rmb seeing something about this somewhere

File: src/main/java/unicash/logic/parser/FilterCommandParser.java (hunk 5 above)
Comment (sp4ce-cowboy) [reply src/main/java/unicash/logic/parser/FilterCommandParser.java:59]: Sorry, the TODO was supposed to be for writing the documentation appendix, to explain "why certain features were done certain way" and this would be along the process of explaining why FilterCommand is limited to one Type prefix only even though that makes it less narrowable (bc rn there are 2 mutually exclusive types, ANDing them would return 0 transactions, but if more types are introduced (i.e. transfers, loans, etc) in AddTransaction, this could change)

File: src/main/java/unicash/logic/parser/UniCashParser.java (hunk 9 above)
Comment (sp4ce-cowboy) [reply src/main/java/unicash/logic/parser/UniCashParser.java:81]: https://github.com/AY2324S1-CS2103-T16-3/tp/pull/130#issuecomment-1785450343

File: src/main/java/unicash/model/transaction/predicates/TransactionContainsAllKeywordsPredicate.java (hunk 11 above)
Comment (sp4ce-cowboy) [reply src/main/java/unicash/model/transaction/predicates/TransactionContainsAllKeywordsPredicate.java:36]: Nice, I didn't catch that, thanks

File: src/main/java/unicash/logic/parser/ClearTransactionsCommandParser.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+public class ClearTransactionsCommandParser implements Parser<ClearTransactionsCommand> {
+
+    /**
+     * Parses {@code userInput} into a {@code ClearTransactionsCommand} and returns it.
+     *
+     * @throws ParseException if {@code userInput} does not conform the expected format
+     */
+    @Override
+    public ClearTransactionsCommand parse(String userInput) throws ParseException {
+        assert userInput != null : "userInput cannot be null";
```
[the remark this answers is not in the document]
Comment (sp4ce-cowboy) [reply src/main/java/unicash/logic/parser/ClearTransactionsCommandParser.java:19]: https://github.com/AY2324S1-CS2103-T16-3/tp/pull/130#discussion_r1376444517

File: src/main/java/unicash/logic/parser/FilterCommandParser.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/logic/parser/FilterCommandParser.java:1]: https://github.com/AY2324S1-CS2103-T16-3/tp/pull/130#issuecomment-1785450343

File: src/main/java/unicash/logic/parser/FilterCommandParser.java [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+            }
+
+        }
+
+        /* Parses the type keyword provided and adds it to the all keywords predicate */
+        if (argMultimap.getValue(PREFIX_TYPE).isPresent()) {
+            Type transactionType = ParserUtil.parseType(
+                    argMultimap.getValue(PREFIX_TYPE).get());
+            filterPredicate.addTypeKeyword(transactionType.toString());
+        }
```
[the remark this answers is not in the document]
Comment (sp4ce-cowboy) [reply src/main/java/unicash/logic/parser/FilterCommandParser.java]: Ok nvm I managed to fix it, I think.

**Comment by woojiahao** [comment]
> help  1 could be flagged as a bug.severityLow or a featureFlaw - where a proposed feature is not "as perfectly implemented" as it can be, even though I agree that it doesn't really make sense to overengineer something like this in the first place.

It seems like the fix here is not to make the input overly restrictive but to make the documentation more comprehensive?

We can always include a line of "There are no arguments required so any arguments included will not be processed" or at least describe clearly the specific usage of the command and ensure that it's obvious. This way, the users can't dispute that we're not handling the case of `clear_transactions today` well since our documentation explicitly mentions that "`clear_transactions` removes all transactions on UniCa$h only. It does not take in any additional parameters or arguments"

> I initially was concerned about overzealous input validation, but that actually applies more to the arguments and their formatting themselves rather than the command words

In this case, this would be considered overzealous input validation, no? Since this parsing does apply to the arguments of the command and has nothing to do with the command word too

**Comment by lipwei1808** [comment]
Just to add on to the discussion, I think going forward, though I think we are already at the end. We should keep all our PR lean. In this case, the problem of **overzealous validation** is hindering the proceeding of the main point of this PR which is filter. The 2 should have been kept apart and anyways 1.9k line changes in 1 PR a week before the submission seems like what the prof was warning us at the start.

**Comment by sp4ce-cowboy** [comment]
> Just to add on to the discussion, I think going forward, though I think we are already at the end. We should keep all our PR lean. In this case, the problem of **overzealous validation** is hindering the proceeding of the main point of this PR which is filter. The 2 should have been kept apart and anyways 1.9k line changes in 1 PR a week before the submission seems like what the prof was warning us at the start.

Just to be clear, the majority of the code is still primarily Filter/Predicate/Parser and associated tests. This has been coupled with @woojiahao's suggestion to add equals methods to all the Class, which I did, and that caused codecoverage to decrease, so I had to add tests for those as well. 1.9k is not changes, it's mostly additions, and if you look at the code the group of affected classes is very narrow i.e. only the Parsers and Tests mostly, I'm just adding more tests (and Filter ofc.) 

Either way, the problem of overzealous input validation and not accounting for legimate user behaviour can cost us quite some marks, so I'm just trying to be safe with adding all the tests.

File: src/main/java/unicash/ui/MainWindow.java [hunk 15]
```diff
@@ -31,12 +24,6 @@
  * a menu bar and space where other JavaFX elements can be placed.
  */
 public class MainWindow extends UiPart<Stage> {
```
Comment (sp4ce-cowboy) [inline src/main/java/unicash/ui/MainWindow.java:26]: Removing User Guide functionality here

File: src/main/resources/view/MainWindow.fxml [hunk 16]
```diff
@@ -33,9 +33,9 @@
             <MenuItem fx:id="helpMenuItem" mnemonicParsing="false"
                       onAction="#handleHelp" text="Help" styleClass="normal-label" />
           </Menu>
-          <Menu mnemonicParsing="false" text="User Guide">
-            <MenuItem fx:id="userGuideMenuItem" text="Open..."
-                      styleClass="normal-label" onAction="#handleOpenUserGuide"/>
+          <Menu mnemonicParsing="false" text="Summary">
+            <MenuItem fx:id="summaryMenuItem" text="Open..."
+                      styleClass="normal-label" onAction="#handleSummary"/>
```
Comment (sp4ce-cowboy) [inline src/main/resources/view/MainWindow.fxml:38]: Swap open UserGuide with open summary here

File: src/test/java/unicash/logic/commands/ResetCommandTest.java [hunk 17]
```diff
@@ -33,4 +38,52 @@ public void execute_nonEmptyUniCash_success() {
...
+    @Test
+    public void sameResetCommand_equalsTrue() {
+        ResetCommand resetCommand = new ResetCommand();
+        assertTrue(resetCommand.equals(resetCommand));
+        assertTrue(resetCommand.equals(new ResetCommand()));
+
+    }
+
+    @Test
+    public void differentCommandTypes_equalsFalse() {
```
Comment (woojiahao) [inline src/test/java/unicash/logic/commands/ResetCommandTest.java:68]: I don't think this is the right naming convention for unit tests, it should be `method_input_output` I believe

File: src/test/java/unicash/logic/parser/GetCommandParserTest.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+    @Test
+    public void differentCommandTypes_equalsFalse() {
+        GetCommandParser getCommandParser = new GetCommandParser();
+        ListCommandParser listCommandParser = new ListCommandParser();
+        assertNotEquals(listCommandParser, getCommandParser);
+        assertFalse(getCommandParser.equals(listCommandParser));
+    }
+
+    @Test
+    public void nullInput_equalsFalse() {
```
Comment (woojiahao) [inline src/test/java/unicash/logic/parser/GetCommandParserTest.java:70]: Same here with unit testing name convention

File: src/test/java/unicash/logic/parser/ListCommandParserTest.java [hunk 19]
```diff
@@ -40,4 +48,35 @@ public void parse_withNullInput_assertionFailure() {
     public void execute_predicateNotNull_assertion() {
         assertDoesNotThrow(() -> new ListCommandParser().parse(" "));
     }
+
+
+    @Test
+    public void sameListCommandParser_equalsTrue() {
```
Comment (woojiahao) [inline src/test/java/unicash/logic/parser/ListCommandParserTest.java:54]: Same here with naming convention

File: src/main/java/unicash/logic/commands/ResetCommand.java
Comment (woojiahao) [inline src/main/java/unicash/logic/commands/ResetCommand.java:1]: Sorry but what's the purpose of this command? It seems like we're adding way too much and we have too little time to find bugs and resolve these

**Review by woojiahao — CHANGES_REQUESTED** [review]
Some comments about the unit testing naming convention. Would also like to ask why we're adding the `ResetCommand` now and what's the difference between the filter and find since filter is just a really large fuzzy search which is a superset of find.

File: src/main/java/unicash/logic/commands/ResetCommand.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/logic/commands/ResetCommand.java:1]: But `ResetCommand` is already in the master branch? Tt was merged [here](https://github.com/AY2324S1-CS2103-T16-3/tp/pull/90). All relevant tests have already been implemented a while back, including tests with the method `getTypicalTransactions()`, so no worries about the bugs, I've already written about 8 tests in `ResetCommandTest` just for this, planning on adding more during v1.4.

File: src/main/java/unicash/logic/commands/ClearTransactionsCommand.java [hunk 20]
```diff
@@ -21,11 +23,36 @@ public class ClearTransactionsCommand extends Command {
...
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof ClearTransactionsCommand)) {
+            return false;
+        }
+
+        return other instanceof ClearTransactionsCommand;
```
[the remark this answers is not in the document]
Comment (sp4ce-cowboy) [reply src/main/java/unicash/logic/commands/ClearTransactionsCommand.java:49]: Fixed, thanks!

File: src/test/java/unicash/logic/commands/ResetCommandTest.java (hunk 17 above)
Comment (sp4ce-cowboy) [reply src/test/java/unicash/logic/commands/ResetCommandTest.java:68]: Fixed.

File: src/test/java/unicash/logic/parser/ListCommandParserTest.java (hunk 19 above)
Comment (sp4ce-cowboy) [reply src/test/java/unicash/logic/parser/ListCommandParserTest.java]: done

File: src/main/java/unicash/logic/parser/GetCommandParser.java [hunk 21]
```diff
@@ -30,5 +31,26 @@ public GetCommand parse(String args) throws ParseException {
...
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof GetCommandParser)) {
+            return false;
+        }
+
+        return other instanceof GetCommandParser;
```
Comment (lipwei1808) [inline src/main/java/unicash/logic/parser/GetCommandParser.java:45]: Same here, but just to check is there a reason we need to ensure quality for the same CommandParser? I don't think we have any requirement for this right

File: src/main/java/unicash/logic/parser/ListCommandParser.java [hunk 22]
```diff
@@ -22,4 +24,27 @@ public ListCommand parse(String userInput) throws ParseException {
...
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof ListCommandParser)) {
+            return false;
+        }
+
+        return other instanceof ListCommandParser;
```
Comment (lipwei1808) [inline src/main/java/unicash/logic/parser/ListCommandParser.java:39]: Same here as the above

**Review by lipwei1808 — COMMENTED** [review]
Overall looks good with the FilterCommand! However, I'm afraid it might be too over-engineered for this. It is good work for sure, but I am not sure if we would want to add in more features as complex (relatively) to our existing codebase when I'm sure we have met the bar relative to other groups in terms of feature implementation. In terms of a group project for a module, I think we might have to weigh in whether it is necessary to add in the new feature for a less than proportionate gain compared to the costs it could bring in terms of exploitation by users. 

I do think it is a really good feature in an application, just maybe not too much for this module where we our graded and with not much runtime left. Maybe the rest can chime in and see if we could maybe lessen the complexity or to merge this in.

But other than that, really nice work man!

**Comment by woojiahao** [comment]
@lipwei1808 Agreed, I think while it's a great feature and I'm sure it works well and as expected, I think it's hard to predict what the testers will do and this late into the project, I think it's safer if we focus our efforts on polishing what we have, and not adding too many extra features

File: src/main/java/unicash/logic/commands/ResetCommand.java [hunk 23]
```diff
@@ -25,10 +26,36 @@ public class ResetCommand extends Command {
...
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof ResetCommand)) {
+            return false;
+        }
+
+        return other instanceof ResetCommand;
```
[the remark this answers is not in the document]
Comment (sp4ce-cowboy) [reply src/main/java/unicash/logic/commands/ResetCommand.java:52]: thanks, fixed!

**Comment by sp4ce-cowboy** [comment]
@lipwei1808 @woojiahao Thank you guys for the review and comments, just some thoughts.

I've been working on Find-Filter for close to two weeks, the initial package of transaction predicates were introduced precisely for this purpose, it just so happened that Find was more easily applicable to that but Filter was always the end goal and as we had discussed in our meetings, I have now completed Filter, including suffiicient testing, and I don't see the rationale for abandoning this once all the work has been put into it. I opened up an earlier [PR](https://github.com/AY2324S1-CS2103-T16-3/tp/pull/122) (git/intellij messed up my work and so I had to re-code it from scratch) so the progress is made visible to everyone and the changes can be tracked easily, precisely because of such potential time constraints.

More importantly, features, bugs, product design and testing are all graded on an individual basis, and Find-Filter is something that I have been working on for quite a while now, Find was implemented with Filter in mind as mentioned [here](https://github.com/AY2324S1-CS2103-T16-3/tp/pull/101#discussion_r1370376068) and most of the design considerations were made with Filter in mind as I had assumed I would be implementing both these commands as we had previously discussed. I planned to elaborate (in the UG) both Find and Filter in a certain way so that Find alone (since it searches for everything) wouldn't be considered a featureFlaw where a certain feature could have been implemented in a better way. In the DG, my rationale for creating separate transaction predicates would also be their versatility, as Find alone could have sufficed with a single unified predicate class. 

I think this late into the project, it doesn't make sense to drastically shift the direction, and this direction as we had discussed in our meetings and PR comment discussions was, at least to my impression, towards the final implementation of Filter.

Nonetheless it is possible to constrain Filter to non-variadic parameters as mentioned [here](https://github.com/AY2324S1-CS2103-T16-3/tp/pull/130#discussion_r1376440192) to make it simpler with a single line of code, but I hope that you guys can take this into consideration and not abandon Filter entirely

File: src/main/java/unicash/logic/commands/FilterCommand.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (jamz903) [reply src/main/java/unicash/logic/commands/FilterCommand.java:40]: While your `Filter` feature is indeed the inverse of `Find`, I feel like this would prove to be confusing for users, who would believe `Find` and `Filter` to be synonymous. These two features typically are represented as separate functions in typical applications (where Filter is extremely specific, down to the T, and find is a generic search, and they can be used in *conjunction*), but filters are typically represented through UI elements. To have the `Filter` and `Find` function both to exist in a CLI application as commands, would make it difficult for users to differentiate and personally I feel that you can merge the two functionalities by adding flairs/tags to the existing `Search` function rather than make a separate one for it. 

For example, `search n/friends -m` could mean to retrieve only results that *completely match*, which would already integrate a big portion of the filter function into search.

**Comment by jamz903** [comment]
Thank you for the immense effort put into this, rubesh! My take on this is to integrate `filter` with the `search` command, as I believe that was what we thought was intended all along, not to add it as a separate command. I agree with what @lipwei1808 and @woojiahao mentioned, and would suggest to lessen the complexity and integrate it with `search`. With a filtered search, it would make the command more robust and we would not completely waste your efforts by stashing this.

**Comment by sp4ce-cowboy** [comment]
> Thank you for the immense effort put into this, rubesh! My take on this is to integrate `filter` with the `search` command, as I believe that was what we thought was intended all along, not to add it as a separate command. I agree with what @lipwei1808 and @woojiahao mentioned, and would suggest to lessen the complexity and integrate it with `search`. With a filtered search, it would make the command more robust and we would not completely waste your efforts by stashing this.

@jamz903 Thanks for replying jamie, I was actually just going to suggest something like this! I was thinking about this for quite a while jn and the more I did the more I realized that actually find and filter can be merged into a single command, because find having no parameters makes it an target for exploitation and also harder to test, so what I'm thinking of doing is maybe remove the filter command but retain its more rigorous logic and incoporate that into find i.e. add parameters to find.

I'll go ahead and remove Filter and try to merge the logic into Find ASAP, thank you guys for all your suggestions and feedback.

**Comment by sp4ce-cowboy** [comment]
Closed, moved to #138
